### ISSUE TITLE
when enabling the new discovery manager, fatal error: concurrent map iteration and map write

### ISSUE THREAD

**Issue by yangfan-njupt** [prometheus/prometheus#9654]
**What did you do?**

run prometheus server v2.31.0 with --enable-feature=new-service-discovery-manager

**What did you see instead? Under which circumstances?**

my prometheus server container restarted 5 times in 3 hours.
by reviewing the log, I found the following fatal error.
```
ts=2021-11-03T07:36:00.098Z caller=main.go:996 level=info msg="Loading configuration file" filename=/etc/config/prometheus.yml
fatal error: concurrent map iteration and map write

goroutine 337 [running]:
runtime.throw({0x2c32669, 0xc000752c00})
        /usr/local/go/src/runtime/panic.go:1198 +0x71 fp=0xc000752b80 sp=0xc000752b50 pc=0x435651
runtime.mapiternext(0xc00099b410)
        /usr/local/go/src/runtime/map.go:858 +0x4eb fp=0xc000752bf0 sp=0xc000752b80 pc=0x40f82b
runtime.mapiterinit(0xc000752db0, 0xc000752cc0, 0xc0648786e0)
        /usr/local/go/src/runtime/map.go:848 +0x236 fp=0xc000752c10 sp=0xc000752bf0 pc=0x40f2f6
github.com/prometheus/prometheus/discovery.(*Manager).allGroups(0xc0001d8ea0)
        /app/discovery/manager.go:376 +0x1e7 fp=0xc000752ef8 sp=0xc000752c10 pc=0xb06f27
github.com/prometheus/prometheus/discovery.(*Manager).sender(0xc0001d8ea0)
        /app/discovery/manager.go:330 +0x178 fp=0xc000752fc8 sp=0xc000752ef8 pc=0xb06618
github.com/prometheus/prometheus/discovery.(*Manager).Run·dwrap·3()
        /app/discovery/manager.go:158 +0x26 fp=0xc000752fe0 sp=0xc000752fc8 pc=0xb04aa6
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1581 +0x1 fp=0xc000752fe8 sp=0xc000752fe0 pc=0x468961
created by github.com/prometheus/prometheus/discovery.(*Manager).Run
        /app/discovery/manager.go:158 +0x5f

```

**Environment**

* System information:

	docker

* Prometheus version:

	v2.31.0

**Comment by yangfan-njupt** [comment]
it looks like forgot to lock when modify manager.targets

https://github.com/prometheus/prometheus/blob/v2.31.0/discovery/manager.go#L220-L222

**Comment by roidelapluie** [comment]
I ended up opening https://github.com/prometheus/prometheus/pull/9656

**Comment by yangfan-njupt** [comment]
@roidelapluie I'd like to work on this, but sorry, I just woke up [ there is time difference here :) ]
maybe next time

**Comment by roidelapluie** [comment]
close by #9656, released in 2.31.1
